src/core/init/duration_updater.ts

```diff
@@ -40,27 +62,139 @@ export default function DurationUpdater(
+ * @param {number | undefined} lastSetDuration
+ * @returns {Observable.<number | null>}
+ */
+function setMediaSourceDuration(
+  mediaSource: MediaSource,
+  manifest: Manifest,
+  lastSetDuration?: number
+): Observable<number | null> {
+  return isMediaSourceOpened$(mediaSource).pipe(
+    switchMap((isMediaSourceOpened) => {
+      if (!isMediaSourceOpened) {
+        return EMPTY;
+      }
+      return whenSourceBuffersEndedUpdates$(mediaSource.sourceBuffers);
+    }),
+    take(1),
+    map(() : number | null => {
+      const maximumPosition = manifest.getMaximumPosition();
+      const { isLive } = manifest;
       // Some targets poorly support setting a very high number for durations.
       // Yet, in live contents, we would prefer setting a value as high as possible
       // to still be able to seek anywhere we want to (even ahead of the Manifest if
       // we want to). As such, we put it at a safe default value of 2^32 excepted
       // when the maximum position is already relatively close to that value, where
       // we authorize exceptionally going over it.
-      const newDuration = !manifest.isLive ? maximumPosition :
-                                             Math.max(Math.pow(2, 32),
-                                                      maximumPosition + YEAR_IN_SECONDS);
+      const newDuration = !isLive ? maximumPosition :
+                                    Math.max(Math.pow(2, 32),
+                                             maximumPosition + YEAR_IN_SECONDS);
 
+      if (mediaSource.duration >= newDuration ||
+          // Even if the MediaSource duration is different than the duration that
+          // we want to set now, the last duration we wanted to set may be the same,
+          // as the MediaSource duration may have been changed by the browser.
+          //
+          // In that case, we do not want to update it.
+          //
+          newDuration === lastSetDuration) {
+        return null;
+      }
       if (isNaN(mediaSource.duration) || !isFinite(mediaSource.duration) ||
-          Math.abs(mediaSource.duration - newDuration) > 0.01)
-      {
+          newDuration - mediaSource.duration > 0.01) {
         log.info("Init: Updating duration", newDuration);
         mediaSource.duration = newDuration;
+        return newDuration;
       }
+      return null;
     }),
-    ignoreElements());
+    catchError((err) => {
+      log.warn("Duration Updater: Can't update duration on the MediaSource.", err);
+      return observableOf(null);
+    })
+  );
+}
+
+/**
+ * Returns an Observable which will emit only when all the SourceBuffers ended
+ * all pending updates.
+ * @param {SourceBufferList} sourceBuffers
+ * @returns {Observable}
+ */
+function whenSourceBuffersEndedUpdates$(
+  sourceBuffers: SourceBufferList
+) : Observable<undefined> {
+  if (sourceBuffers.length === 0) {
+    return observableOf(undefined);
+  }
+  const updatingSourceBuffers$: Array<Observable<boolean>> = [];
```

Comment:
Could this naming bring misunderstanding as `updatingSourceBuffers$` is not an observable but rather an Array?

Reply:
Yes

Reply:
Renamed to `sourceBufferUpdatingStatuses`